File: mvikotlin-timetravel-client/plugin-idea/src/main/java/com/arkivanov/mvikotlin/plugin/idea/timetravel/TimeTravelToolbar.java
package com.arkivanov.mvikotlin.plugin.idea.timetravel;

import com.arkivanov.mvikotlin.timetravel.client.internal.TimeTravelClientView;
import com.intellij.icons.AllIcons;
import com.intellij.openapi.actionSystem.ActionManager;
import com.intellij.openapi.actionSystem.ActionPlaces;
import com.intellij.openapi.actionSystem.ActionToolbar;
import com.intellij.openapi.actionSystem.AnAction;
import com.intellij.openapi.actionSystem.AnActionEvent;
import com.intellij.openapi.actionSystem.DefaultActionGroup;
import com.intellij.openapi.actionSystem.Presentation;

import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.function.Consumer;

import javax.swing.Icon;
import javax.swing.JComponent;

public class TimeTravelToolbar {

    @NotNull
    private final Listener listener;
    @NotNull
    private final ActionToolbar toolbar;

    @NotNull
    private TimeTravelClientView.Model.Buttons buttons =
            new TimeTravelClientView.Model.Buttons(
                    false,
                    false,
                    false,
                    false,
                    false,
                    false,
                    false,
                    false,
                    false,
                    false,
                    false,
                    false
            );

    public TimeTravelToolbar(@NotNull Listener listener) {
        this.listener = listener;

        this.toolbar = ActionManager.getInstance().createActionToolbar(ActionPlaces.COMMANDER_TOOLBAR, actionGroup(), true);
    }

    @NotNull
    public JComponent getComponent() {
        return toolbar.getComponent();
    }

    public void render(@NotNull TimeTravelClientView.Model.Buttons buttons) {
        this.buttons = buttons;
        toolbar.updateActionsImmediately();
    }

    @NotNull
    private DefaultActionGroup actionGroup() {
        final DefaultActionGroup group = new DefaultActionGroup();
        group.addAll(connectAction(), disconnectAction());
        group.addSeparator();
        group.addAll(
                startRecordingAction(),
                stopRecordingAction(),
                moveToStartAction(),
                stepBackwardAction(),
                stepForwardAction(),
                moveToEndAction(),
                cancelAction()
        );
        group.addSeparator();
        group.add(debugAction());
        group.addSeparator();
        group.add(exportAction());
        group.add(importAction());

        return group;
    }

    @NotNull
    private AnAction connectAction() {
        return anAction(
                "Connect",
                AllIcons.Debugger.AttachToProcess,
                (event) -> event.getPresentation().setEnabled(buttons.isConnectEnabled()),
                listener::onConnect
        );
    }

    @NotNull
    private AnAction disconnectAction() {
        return anAction(
                "Disconnect",
                AllIcons.Debugger.Db_invalid_breakpoint,
                (event) -> event.getPresentation().setEnabled(buttons.isDisconnectEnabled()),
                listener::onDisconnect
        );
    }

    @NotNull
    private AnAction startRecordingAction() {
        return anAction(
                "Start recording",
                AllIcons.Debugger.Db_set_breakpoint,
                (event) -> event.getPresentation().setEnabled(buttons.isStartRecordingEnabled()),
                listener::onStartRecording
        );
    }

    @NotNull
    private AnAction stopRecordingAction() {
        return anAction(
                "Stop recording",
                AllIcons.Actions.Suspend,
                (event) -> event.getPresentation().setEnabled(buttons.isStopRecordingEnabled()),
                listener::onStopRecording
        );
    }

    @NotNull
    private AnAction moveToStartAction() {
        return anAction(
                "Move to start",
                AllIcons.Actions.Play_first,
                (event) -> event.getPresentation().setEnabled(buttons.isMoveToStartEnabled()),
                listener::onMoveToStart
        );
    }

    @NotNull
    private AnAction stepBackwardAction() {
        return anAction(
                "Step backward",
                AllIcons.Actions.Play_back,
                (event) -> event.getPresentation().setEnabled(buttons.isMoveToEndEnabled()),
                listener::onStepBackward
        );
    }

    @NotNull
    private AnAction stepForwardAction() {
        return anAction(
                "Step forward",
                AllIcons.Actions.Play_forward,
                (event) -> event.getPresentation().setEnabled(buttons.isStepBackwardEnabled()),
                listener::onStepForward
        );
    }

    @NotNull
    private AnAction moveToEndAction() {
        return anAction(
                "Move to end",
                AllIcons.Actions.Play_last,
                (event) -> event.getPresentation().setEnabled(buttons.isMoveToEndEnabled()),
                listener::onMoveToEnd
        );
    }

    @NotNull
    private AnAction cancelAction() {
        return anAction(
                "Cancel",
                AllIcons.Actions.Cancel,
                (event) -> event.getPresentation().setEnabled(buttons.isCancelEnabled()),
                listener::onCancel
        );
    }

    @NotNull
    private AnAction debugAction() {
        return anAction(
                "Debug selected event",
                AllIcons.Actions.StartDebugger,
                (event) -> event.getPresentation().setEnabled(buttons.isDebugEventEnabled()),
                listener::onDebug
        );
    }

    @NotNull
    private AnAction exportAction() {
        return anAction(
                "Export events",
                AllIcons.ToolbarDecorator.Export,
                (event) -> event.getPresentation().setEnabled(buttons.isExportEventsEnabled()),
                listener::onExport
        );
    }

    @NotNull
    private AnAction importAction() {
        return anAction(
                "Import events",
                AllIcons.ToolbarDecorator.Import,
                (event) -> event.getPresentation().setEnabled(buttons.isImportEventsEnabled()),
                listener::onImport
        );
    }

    @NotNull
    private AnAction anAction(
            @Nullable String text,
            @Nullable Icon icon,
            @Nullable Consumer<AnActionEvent> onUpdate,
            @NotNull Runnable onAction
    ) {
        final AnAction action =
                new AnAction() {
                    @Override
                    public void actionPerformed(@NotNull AnActionEvent event) {
                        onAction.run();
                    }

                    @Override
                    public void update(@NotNull AnActionEvent event) {
                        if (onUpdate != null) {
                            onUpdate.accept(event);
                        }
                    }
                };

        final Presentation presentation = action.getTemplatePresentation();
        presentation.setText(text);
        presentation.setIcon(icon);

        return action;
    }

    interface Listener {
        void onConnect();

        void onDisconnect();

        void onStartRecording();

        void onStopRecording();

        void onMoveToStart();

        void onStepBackward();

        void onStepForward();

        void onMoveToEnd();

        void onCancel();

        void onDebug();

        void onExport();

        void onImport();
    }
}

File: mvikotlin-timetravel-client/plugin-idea/src/main/java/com/arkivanov/mvikotlin/plugin/idea/timetravel/TimeTravelToolbar.kt
package com.arkivanov.mvikotlin.plugin.idea.timetravel

import com.arkivanov.mvikotlin.timetravel.client.internal.client.TimeTravelClient.Model.Buttons
import com.intellij.icons.AllIcons
import com.intellij.openapi.actionSystem.ActionManager
import com.intellij.openapi.actionSystem.ActionPlaces
import com.intellij.openapi.actionSystem.ActionToolbar
import com.intellij.openapi.actionSystem.AnAction
import com.intellij.openapi.actionSystem.DefaultActionGroup
import javax.swing.JComponent

class TimeTravelToolbar(private val listener: Listener) {

    private val toolbar: ActionToolbar =
        ActionManager
            .getInstance()
            .createActionToolbar(ActionPlaces.COMMANDER_TOOLBAR, actionGroup(), true)

    private var buttons: Buttons =
        Buttons(
            isConnectEnabled = false,
            isDisconnectEnabled = false,
            isStartRecordingEnabled = false,
            isStopRecordingEnabled = false,
            isMoveToStartEnabled = false,
            isStepBackwardEnabled = false,
            isStepForwardEnabled = false,
            isMoveToEndEnabled = false,
            isCancelEnabled = false,
            isDebugEventEnabled = false,
            isExportEventsEnabled = false,
            isImportEventsEnabled = false
        )

    val component: JComponent get() = toolbar.component

    fun render(buttons: Buttons) {
        this.buttons = buttons
        toolbar.updateActionsImmediately()
    }

    private fun actionGroup(): DefaultActionGroup =
        DefaultActionGroup().apply {
            addAll(connectAction(), disconnectAction())
            addSeparator()
            addAll(
                startRecordingAction(),
                stopRecordingAction(),
                moveToStartAction(),
                stepBackwardAction(),
                stepForwardAction(),
                moveToEndAction(),
                cancelAction()
            )
            addSeparator()
            add(debugAction())
            addSeparator()
            add(exportAction())
            add(importAction())
        }

    private fun connectAction(): AnAction =
        anAction(
            text = "Connect",
            icon = AllIcons.Debugger.AttachToProcess,
            onUpdate = { it.presentation.isEnabled = buttons.isConnectEnabled },
            onAction = listener::onConnect
        )

    private fun disconnectAction(): AnAction =
        anAction(
            text = "Disconnect",
            icon = AllIcons.Debugger.Db_invalid_breakpoint,
            onUpdate = { it.presentation.isEnabled = buttons.isDisconnectEnabled },
            onAction = listener::onDisconnect
        )

    private fun startRecordingAction(): AnAction =
        anAction(
            text = "Start recording",
            icon = AllIcons.Debugger.Db_set_breakpoint,
            onUpdate = { it.presentation.isEnabled = buttons.isStartRecordingEnabled },
            onAction = listener::onStartRecording
        )

    private fun stopRecordingAction(): AnAction =
        anAction(
            text = "Stop recording",
            icon = AllIcons.Actions.Suspend,
            onUpdate = { it.presentation.isEnabled = buttons.isStopRecordingEnabled },
            onAction = listener::onStopRecording
        )

    private fun moveToStartAction(): AnAction =
        anAction(
            text = "Move to start",
            icon = AllIcons.Actions.Play_first,
            onUpdate = { it.presentation.isEnabled = buttons.isMoveToStartEnabled },
            onAction = listener::onMoveToStart
        )

    private fun stepBackwardAction(): AnAction =
        anAction(
            text = "Step backward",
            icon = AllIcons.Actions.Play_back,
            onUpdate = { it.presentation.isEnabled = buttons.isMoveToEndEnabled },
            onAction = listener::onStepBackward
        )

    private fun stepForwardAction(): AnAction =
        anAction(
            text = "Step forward",
            icon = AllIcons.Actions.Play_forward,
            onUpdate = { it.presentation.isEnabled = buttons.isStepBackwardEnabled },
            onAction = listener::onStepForward
        )

    private fun moveToEndAction(): AnAction =
        anAction(
            text = "Move to end",
            icon = AllIcons.Actions.Play_last,
            onUpdate = { it.presentation.isEnabled = buttons.isMoveToEndEnabled },
            onAction = listener::onMoveToEnd
        )

    private fun cancelAction(): AnAction =
        anAction(
            text = "Cancel",
            icon = AllIcons.Actions.Cancel,
            onUpdate = { it.presentation.isEnabled = buttons.isCancelEnabled },
            onAction = listener::onCancel
        )

    private fun debugAction(): AnAction =
        anAction(
            text = "Debug selected event",
            icon = AllIcons.Actions.StartDebugger,
            onUpdate = { it.presentation.isEnabled = buttons.isDebugEventEnabled },
            onAction = listener::onDebug
        )

    private fun exportAction(): AnAction =
        anAction(
            text = "Export events",
            icon = AllIcons.ToolbarDecorator.Export,
            onUpdate = { it.presentation.isEnabled = buttons.isExportEventsEnabled },
            onAction = listener::onExport
        )

    private fun importAction(): AnAction =
        anAction(
            text = "Import events",
            icon = AllIcons.ToolbarDecorator.Import,
            onUpdate = {
                it.presentation.isEnabled = buttons.isImportEventsEnabled
            },
            onAction = listener::onImport
        )


    interface Listener {
        fun onConnect()
        fun onDisconnect()
        fun onStartRecording()
        fun onStopRecording()
        fun onMoveToStart()
        fun onStepBackward()
        fun onStepForward()
        fun onMoveToEnd()
        fun onCancel()
        fun onDebug()
        fun onExport()
        fun onImport()
    }
}
